stablecoin-v3/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+            user_payments.push(user_payment)
+        }
+
+        self.send().direct_multi(&caller, &user_payments, &[]);
+
+        payment_value_denominated
+    }
+
+    #[payable("*")]
+    #[endpoint(swapStablecoin)]
+    fn swap_stablecoin(&self, amount_out_min: BigUint) {
+        require!(self.can_swap(), ERROR_SWAP_NOT_ENABLED);
+
+        let (amount_in, token_in) = self.call_value().payment_token_pair();
+        let collateral_token_id = self.collateral_token_id().get();
+        let stablecoin_token_id = self.stablecoin().get_token_id();
+
+        require!(
+            token_in == collateral_token_id || token_in == stablecoin_token_id,
+            ERROR_BAD_PAYMENT_TOKENS
+        );
+
+        // Sets the swap type (stablecoin buy/sell)
+        let mut stablecoin_buy = true;
+        let mut token_out = stablecoin_token_id.clone();
+        if token_in == stablecoin_token_id {
+            stablecoin_buy = false;
+            token_out = collateral_token_id.clone();
+        }
+
+        let caller = self.blockchain().get_caller();
+
+        // Replenish delta at start of each block
+        self.update_pool_delta();
+
+        // Compute swap
+
+        // Oracle prices
+        let collateral_price = self.get_exchange_rate(&collateral_token_id, &stablecoin_token_id);
+        let stablecoin_price = self.get_exchange_rate(&stablecoin_token_id, &stablecoin_token_id);
+
+        let base_pool = self.base_pool().get();
+        let min_swap_spread = BigUint::from(self.spread_fee_min_percent().get());
+        let cp = &base_pool * &base_pool;
+        let pool_delta = self.pool_delta().get();
+        let median_pool_delta = BigUint::from(MEDIAN_POOL_DELTA);
+
+        // stablecoin_pool = base_pool + pool_delta;
+        let mut stablecoin_pool = base_pool.clone();
+        if pool_delta > median_pool_delta {
+            stablecoin_pool += &pool_delta - &median_pool_delta;
+        } else if median_pool_delta > pool_delta {
+            stablecoin_pool -= &median_pool_delta - &pool_delta;
+        }
+
+        let collateral_pool = &cp / &stablecoin_pool;
+
+        let offer_rate;
+        let demand_rate;
+        let offer_pool;
+        let demand_pool;
+        if stablecoin_buy {
+            offer_rate = collateral_price.clone();
+            demand_rate = stablecoin_price.clone();
+            offer_pool = collateral_pool.clone();
+            demand_pool = stablecoin_pool.clone();
+        } else {
+            offer_rate = stablecoin_price.clone();
+            demand_rate = collateral_price.clone();
+            offer_pool = stablecoin_pool.clone();
+            demand_pool = collateral_pool.clone();
+        }
+
+        // Calculate optimal value of amount_out
+        let amount_out_optimal = &amount_in * &offer_rate / &demand_rate;
+
+        // Constant-product based swap amount
+        let demand_base_amount = &demand_pool - &(&cp / &(&offer_pool + &amount_out_optimal));
```

Comment:
We'll have to look over those formulas at some point, I'm not entirely sure they're correct.